### PR TITLE
Compile-time checked State

### PR DESCRIPTION
**Pull request by Victor-N-Suadicani**
Adjusts how state is managed in the App. Now, a state type `S` must be given to the app at construction in `App::new`. Handlers can now extract state types that implement `From<&S>`. `S` will be used to construct the state via cloning. This makes it so that the `State<T>` extractor is now infallible.

### CODE REVIEW

**Comment by gorm-issuu** [comment]
Why are tests passing? The doc tests cannot possibly be correct :thinking:

```rs
//! #[tokio::main]
//! async fn main() -> kanin::Result<()> {
//!     App::new()
//!         .handler("my_routing_key", echo)
//!         .run("amqp://localhost")
//!         .await
//! }
//! ```
```

**Comment by Victor-N-Suadicani** [comment]
> Why are tests passing? The doc tests cannot possibly be correct 🤔

I think they weren't green but not sure what code you're referring to now - anyways, green now :D

**Comment by gorm-issuu** [comment]
> I think they weren't green but not sure what code you're referring to now - anyways, green now :D

Actually I think I was referring to outdated code. My bad. But now I found [the minimal example](https://github.com/issuu/kanin/blob/main/minimal_example/src/main.rs#L14) to not be up to date.

In general I am missing some documentation on this somewhere.

**Comment by Victor-N-Suadicani** [comment]
You're looking at main instead of this branch. The minimal example is mentioned in the README.

File: kanin/src/extract/state.rs [hunk 1]
```diff
@@ -1,15 +1,18 @@
...
 
-/// `State` is an extractor helper struct that allows you to extract app state
-/// that has previously been added to the app through a call to [`crate::App::state`]
+/// `State` is an extractor helper struct that allows you to extract app state from the state type added in `App::new`.
+///
+/// This implements `Deref` and `DerefMut` to the inner type, but the inner type is also public so you can also just take ownership if you prefer.
 ///
-/// This implements `Deref` and `DerefMut` to the inner type.
+/// Any type that implements `From<&S>` where `S` is the app state given in `App::new` can be extracted via this type.
+/// These `From` implementations can be derived on a struct via `kanin::AppState`.
```
Comment (gorm-issuu) [inline kanin/src/extract/state.rs:15]:
```suggestion
/// These `From` implementations can be derived on a struct via `kanin::AppState`.
///
/// # Example
/// ```
/// # struct AppState;
/// async fn my_handler(State(app_state): State<AppState>) {
///     // ...
/// }
/// ```
```

It took me longer than I care to admit to realize that this is just the extractor. I think documentation could be stronger with some examples.

File: kanin/src/extract/state.rs [hunk 2]
```diff
@@ -21,16 +24,14 @@ impl<T: Clone> Clone for State<T> {
 
 /// Extract implementation for app state.
 #[async_trait]
-impl<T: Clone + Send + Sync + 'static> Extract for State<T> {
-    type Error = HandlerError;
+impl<S, T> Extract<S> for State<T>
```
Comment (gorm-issuu) [inline kanin/src/extract/state.rs:27]: Same that we still need the `async_trait`

File: kanin/src/lib.rs [hunk 3]
```diff
@@ -67,7 +67,7 @@
 //!
 //! #[tokio::main]
 //! async fn main() -> kanin::Result<()> {
-//!     App::new()
+//!     App::new(())
```
Comment (gorm-issuu) [inline kanin/src/lib.rs:70]: I don't know if this is the place, but I really think some documentation of how to use state, including the derive, is missing.

**Review by gorm-issuu — CHANGES_REQUESTED** [review]
Overall it looks awesome. Really happy to get this in. My only request is that we have an actual example somewhere for people to have a look at.

Another thing I just randomly noticed, is that `cargo doc` builds with 4 warnings. All of them are completely unrelated to this PR, but maybe you want to fix them before next release :)

File: kanin/src/lib.rs (hunk 3 above)
Comment (Victor-N-Suadicani) [reply kanin/src/lib.rs]: Added an example here too.

**Review by gorm-issuu — APPROVED** [review]

**Pull request merged**